### ISSUE TITLE
RAII for database connections

### ISSUE THREAD

**Issue by falkmielke** [inbo/inbodb#59]
Coming from a different programming language and accessing my first INBO database today :tada:, I noticed that vignettes still contain manual `dbDisconnect()` calls. 
There are ways of "pre-registering" a disconnect call, the concept is called RAII (Resource Acquisition Is Initialization). I quickly searched which options are available in `R`.

I found [this blog post](https://shrektan.com/post/2019/07/26/create-a-database-connection-that-can-be-disconnected-automatically) where the author builds up a "finalizer" using `on.exit()` ([docs](https://rdrr.io/r/base/on.exit.html)).
We should also check the `later` and `withr` packages, for other examples in similar use cases (found via [this SO post](https://stackoverflow.com/a/28300714), [later package](https://cran.r-project.org/web/packages/later/index.html)).
Without thorough comparison and expertise, I would guess the simplest way to be `withr::with_db_connection()` ([docs](https://withr.r-lib.org/reference/with_db_connection.html)).

The benefit would be that we don't burden our databases with open connections.
I can imagine that especially in non-linear work (e.g. Rmd chunks) connection count could staple up involuntarily; might be worth checking whether this is an actual burden.
At any rate, closing the connection is good practice, and ensuring the disconnect _a priori_ seems worthwile.

**Comment by falkmielke** [comment]
Small update, and possible solution:
Being busy on [my own database playground](https://github.com/inbo/n2khab-mne-monitoring/tree/database_initialization/900_database_organization), I found `reg.finalizer()`, which seems a viable solution to this issue and is actually base R!

Here  an example:

```{r}

connect_database_configfile <- function(config_filepath) {

  # read connection info from a config file
  config <- configr::read.config(file = config_filepath)[[1]] # only connects with first config block


  # connect to database
  database_connection <- DBI::dbConnect(
    RPostgres::Postgres(),
    dbname = config$database,
    host = config$server,
    port = config$port,
    user = config$user,
    password = config$password
  )

  # store a label for verbose disconnection at exit
  db_label <- sprintf("%s@%s/%s", config$user, config$server, config$database)

  # remove the config: we do not want to expose credentials further
  # down in this notebook
  rm(config)

  # register disconnect for finalization
  # https://stackoverflow.com/a/41179916
  reg.finalizer(
    .GlobalEnv,
    function(e){
      DBI::dbDisconnect(database_connection)
      message(sprintf("database %s disconnected.", db_label))
    },
    onexit = TRUE
  )
  
  return(invisible(database_connection))
}

# usage:
config_filepath <- file.path("./inbopostgis_server.conf")
db_connection <- connect_database_configfile(config_filepath)

```

As a bonus, this uses a config file to store the connection credentials. Potential improvement would be to prompt or encrypt the password.

```{toml}
    [test]
    server = localhost
    port = 5439
    user = test
    database = playground
    password = <the password you entered IN PLAIN TEXT>
```

**Comment by ElsLommelen** [comment]
Before testing it in practice, first some questions about your solution:
- If I understand well, the connection only closes if the R session ends, so it stays open for hours if a user is busy developing a script or running a long calculation after loading some data from a database?  Why keeping the connection open for so long if a simple `dbDisconnect()` command can close the connection immediately after loading the data?
- I sometimes get messages that a connection is closed if I forget to close it myself, so I wonder to what extent DBI or odbc already has a similar functionality implemented.  And if not, wouldn't it be more useful to propose your solution there, so much more users can benefit from it?  inbodb is mainly developed to access our own databases and as a 'warehouse' with easy access to often used data (no queries or knowledge of the database structure needed), we do not aim to improve technical issues from R packages that we are using to make the connection (unless adaptations for our typical working environment are needed).
- inbodb typically uses the computer user account credentials which makes the current system very easy for INBO users (no need to add credentials for the connection).  Saving credentials in a config file seems to complicate things a lot each time the user changes the password, so honestly I don't see any advantages in saving the credentials locally?

**Comment by falkmielke** [comment]
> If I understand well, the connection only closes if the R session ends, so it stays open for hours if a user is busy developing a script or running a long calculation after loading some data from a database?  Why keeping the connection open for so long if a simple `dbDisconnect()` command can close the connection immediately after loading the data?

Indeed, if the connection has been closed before, no need to execute the finalizer. It might be good to check whether the connection is still active, unless `dbDisconnect` is indifferent to activity.

 
> I sometimes get messages that a connection is closed if I forget to close it myself, so I wonder to what extent DBI or odbc already has a similar functionality implemented.  And if not, wouldn't it be more useful to propose your solution there, so much more users can benefit from it?  inbodb is mainly developed to access our own databases and as a 'warehouse' with easy access to often used data (no queries or knowledge of the database structure needed), we do not aim to improve technical issues from R packages that we are using to make the connection (unless adaptations for our typical working environment are needed).

If DBI or odbc have a similar solution (which I do not know), this whole issue is obsolete: it stemmed from the observation that we do not consistently close our connections (and the good practice to already schedule the disconnect upon initial connection).
But chances are that the disconnect messages you get are network instabilities on either side? (I haven't seen those, but also not worked long with it).


> inbodb typically uses the computer user account credentials which makes the current system very easy for INBO users (no need to add credentials for the connection).  Saving credentials in a config file seems to complicate things a lot each time the user changes the password, so honestly I don't see any advantages in saving the credentials locally?

True, and agreed. Saving credentials is useful for me since I temporarily work outside the scope, and switch around with admin- and user roles a lot. In my case, text config files are handy for switching (and can be stored in an [encrypted space](https://dyne.org/tomb/)).
I found [the `keyring` package](https://keyring.r-lib.org/) to prompt passwords, and [`configr`](https://github.com/Miachol/configr) is great to manage the configs.[^1]

[^1]: *edit: I would intuitively find it bad practice to have the same credentials for a database and our systems. This extends our systems attack surface.*
___

As I see it, and as I initially mentioned, there is probably little benefit of being strict with disconnecting, `inbodb` is working fine as is, and solving this issue might take some (uncertain) effort. 
You might thus well decide to close this issue!
